### ISSUE TITLE
Cart cache problem

### ISSUE THREAD

**Issue by erikdemarco** [woocommerce/storefront#1680]
I use cloudflare and set to cache my product page. The cart contents show other people's cart before updated to current visitor cart contents. The updating takes split seconds, but still it looks unprofessional. Why other people can see other people contents?
I think it not just cloudflare, it will be the same with other cache plugin.

To fix:
Cart widget (storefront_cart_link), on first load the contents of the cart should be empty, before ajax takes care of showing the updated contents.

**Comment by nerrad** [comment]
Hi there @erikdemarco, thanks for taking the time to leave your report here.

I'm having a bit of trouble understanding how this is related to Storefront? I think you are saying that when you cache your product page using Cloudflare that cart contents (for the mini cart widget?) are showing from the cached content before the cart is updated for the current user? You refer to the [`storefront_cart_link` function](https://github.com/woocommerce/storefront/blob/e890ea8b416088f6974e84afc98fdab016049c75/inc/woocommerce/storefront-woocommerce-template-functions.php#L86) but as far as I can tell that's only for updating certain fragments within the mini cart widget html and not all the content.

So it appears this may be an issue you'd need to raise in the [WooCommerce core](https://github.com/woocommerce/woocommerce/issues) repository. However, given you have caching configured for your site, this also may be related to how the caching is configured.

I'm going to close this issue for now as there's nothing immediately actionable on our end, but if necessary and there's some clearer indications this is a Storefront issue, we will reopen.

**Comment by erikdemarco** [comment]
@nerrad Hi thank you for having time to review this issue.

Here's ELI5 version of the issue:

There is 2 visitors visiting my store, lets say VisitorA and VisitorB.

1. VisitorA visit my site first, and add cart 1 productA. 

2. Then storefront render the static html for that minicart part using storefront_cart_link()
in that static minicart html it contains VisitorA cart. (judging by WC()->cart->get_cart_subtotal() inside storefront_cart_link() )

3. My server saved this static html page as a cache so next visitor who visited this page doesnt need to render this page again.

4.  The static page get shown to VisitorA, then there is javascript script updating minicart once again for the current session of the visitor's cart.

Until this part nothing is wrong. Then come VisitorB, Here's the breakdown:

1. VisitorB visit my site first, and add cart 1 productA. 

2. Then my server show the cached version of this page (which have visitorA cart in it)

3. Then this static page get shown to VisitorB. He will see and think 'Why my cart total is different from what i added to cart'?

4. And the javascript runs to update the minicart widget once again and showing the current session for this user. 

5. VisitorB will now be shown his own cart total. He will think "Now this is my own cart, But this site looks unprofessional, how can I see other people's cart totla even for just split second?".


To fix this problem, storefront needs to separate storefront_cart_link() function for ajax and for page rendering.
For page rendering it must show empty cart. 
For ajax rendering it must show the curent session's cart.

**Comment by nerrad** [comment]
Thanks, @erikdemarco - your explanation confirms what I thought you were trying to convey. However, given your explanation I still don't think this is a Storefront issue as the actual code for the mini cart generation lives in WooCommerce core.

`storefront_cart_link` is called by [`storefront_cart_link_fragment`](https://github.com/woocommerce/storefront/blob/trunk/inc/woocommerce/storefront-woocommerce-template-functions.php#L63) which in turn is a callback registered to the [`woocommerce_add_to_cart_fragments`](https://github.com/woocommerce/storefront/blob/e890ea8b416088f6974e84afc98fdab016049c75/inc/woocommerce/storefront-woocommerce-template-hooks.php#L92) filter. This filter is applied in [WooCommerce core here](https://github.com/woocommerce/woocommerce/blob/bee633cef7fcc7eabd6f24fa3bc5b8e04133e674/includes/class-wc-ajax.php#L192-L200).
